### PR TITLE
RAG app example

### PR DESCRIPTION
**Pull request by heyjustinai**
# What does this PR do?

Creating a E2E RAG example that is able to do retrieval on documents and answer user questions. Components included:

Inference (with llama-stack)
Memory (with llama-stack)
Agent (with llama-stack)
Frontend (with Gradio)

## Feature/Issue validation/testing/test plan

image

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/meta-llama/llama-stack-apps/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: examples/agents/rag_with_memory_bank.py [hunk 1]
```diff
@@ -36,17 +36,31 @@ async def run_main(host: str, port: int, disable_safety: bool = False):
 
     client = LlamaStackClient(base_url=f"http://{host}:{port}")
     providers = client.providers.list()
+
+    model_name = "Llama3.2-3B-Instruct"
```
Comment (ashwinb) [inline examples/agents/rag_with_memory_bank.py:40]: are these spurious changes?

File: examples/E2E-RAG-App/01_ingest_files.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
+import argparse
```
Comment (ashwinb) [inline examples/E2E-RAG-App/01_ingest_files.py:1]: I think we should simply call this directory `DocQA` with the organization:

- DocQA
  - app.py
  - README.md
  - scripts/
  - data/

Could we also avoid prefixing files `01_` or `02_` etc.

File: examples/E2E-RAG-App/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+### Prerequisite:
+
+Install docker: Check [this doc for Mac](https://docs.docker.com/desktop/setup/install/mac-install/), [this doc for Windows](https://docs.docker.com/desktop/setup/install/windows-install/) and this [instruction for Linux](https://docs.docker.com/engine/install/).
+
+For Mac and Windows users, you need to start the Docker app manually after installation.
+
+### How to run the pipeline:
+
+![RAG_workflow](./RAG_workflow.jpg)
```
Comment (ashwinb) [inline examples/E2E-RAG-App/README.md:15]: any chance we can simplify this diagram a _lot_? actually, I think a simpler inline Mermaid diagram which shows just the basic high-level flow would be more useful. docker, etc. should be completely avoided.

File: examples/E2E-RAG-App/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+1. We have main config `RAG_service.json` inside of the docker folder, please change `model_name` and `document_path` accordingly, for example:
+
+```yaml
+{
+  "model_name": "llama3.2:1b-instruct-fp16",
+  "document_path": "${HOME}/work/llama-stack-apps/examples/E2E-RAG-App/example_data"
+}
+```
+
+2. Inside of docker folder, `run_RAG.sh` is the main script that can create `.env` file for compose.yaml and then actually start the `docker compose` process to launch all the pipelines in our dockers. `compose.yaml` is the main docker yaml that specifies all the mount option and docker configs, change the mounts if needed.
```
Comment (ashwinb) [inline examples/E2E-RAG-App/README.md:28]: these details are unnecessary I think. these scripts are very simple and self-documenting in a way.

File: examples/E2E-RAG-App/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+```bash
+docker exec -it docker-ollama-1 bash
+ollama ps
+```
+
+Check more about Ollama instruction [here](https://github.com/ollama/ollama)
+
+4. ChromaDB docker will also start. This docker will host the chroma database that can interact with llama-stack.
+
+5. Lastly, Llama-stack docker will start. The `llama_stack_start.sh` control the docker startup behavior, change it if needed. It will first run the ingestion pipeline to convert all the documents into MarkDown files. Then, it will run llama-stack server based on the  `llama_stack_run.yaml` config. Once the server is ready, then it will run the `gradio_interface.py` which will insert document chunks into memory_bank and start the UI for user interaction.
```
Comment (ashwinb) [inline examples/E2E-RAG-App/README.md:52]: we have scripts within scripts, is it possible to inline some of them perhaps?

File: examples/E2E-RAG-App/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+6. `gradio_interface.py` will show a public link. You can access the gradio UI by putting this link to the browser. Then you can start your chat in the gradio web page.
```
Comment (ashwinb) [inline examples/E2E-RAG-App/README.md:54]: this should be renamed `app.py` as commented earlier

File: examples/E2E-RAG-App/data/eval/eval.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
+import asyncio
```
Comment (ashwinb) [inline examples/E2E-RAG-App/data/eval/eval.py:1]: wow, having evals here is awesome! ❤️

File: examples/E2E-RAG-App/data/eval/eval.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    Faithfulness,
+    SemanticSimilarity,
+)
+from termcolor import cprint
+from tqdm import tqdm
+
+# Initialization
+load_dotenv()
+embedding_function = SentenceTransformerEmbeddingFunction(
+    model_name="all-mpnet-base-v2"
```
Comment (ashwinb) [inline examples/E2E-RAG-App/data/eval/eval.py:31]: isn't this a different model than the one we use inside llama stack?

File: examples/E2E-RAG-App/gradio_interface.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+                    "embedding_model": "all-MiniLM-L6-v2",
+                    "chunk_size_in_tokens": 100,
+                    "overlap_size_in_tokens": 10,
+                },
+                provider_id=provider_id,
+            )
+            await self.load_documents()
+            print(f"Memory bank registered.")
+
+    async def load_documents(self):
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:73]: a little confused -- I thought the workflow was that we _pre_ ingest documents and then the chat app merely works with such a pre-ingested, already set up memory bank. because otherwise I don't understand why we needed `01_ingest_files.py`?

File: examples/E2E-RAG-App/gradio_interface.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+        if "1b" in MODEL_NAME:
+            model_name = "Llama3.2-1B-Instruct"
+        elif "3b" in MODEL_NAME:
+            model_name = "Llama3.2-3B-Instruct"
+        elif "8b" in MODEL_NAME:
+            model_name = "Llama3.1-8B-Instruct"
+        else:
+            model_name = MODEL_NAME
+
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:107]: we should not do any such thing since this is code that developers will see and get confused. let's just use the full HF repo names since they are now supported by the stack

File: examples/E2E-RAG-App/gradio_interface.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+            model=model_name,
+            instructions="You are a helpful assistant that can answer questions based on provided documents. Return your answer short and concise, less than 50 words.",
+            sampling_params={"strategy": "greedy", "temperature": 1.0, "top_p": 0.9},
+            tools=[
+                {
+                    "type": "memory",
+                    "memory_bank_configs": [
+                        {"bank_id": self.memory_bank_id, "type": "vector"}
+                    ],
+                    "query_generator_config": {"type": "default", "sep": " "},
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:118]: is this parameter needed? is the default value OK? it is so confusing! (our bad, of course)

File: examples/E2E-RAG-App/gradio_interface.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+                    "max_tokens_in_context": 300,
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:119]: this parameter is also hard to understand for a developer. just a note for us.

File: examples/E2E-RAG-App/gradio_interface.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+                    "max_chunks": 5,
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:120]: I think we should just elide this and leave it to default. because we don't want people to be thinking about all these pieces when they first look at the stack (and even later if we do a good job)

File: examples/E2E-RAG-App/gradio_interface.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+                }
+            ],
+            tool_choice="auto",
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:123]: let's lean on default values here to simplify code

File: examples/E2E-RAG-App/gradio_interface.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        history.append([message, ""])
+
+        output_queue = Queue()
+
+        def run_async():
+            async def async_process():
+                if self.agent is None:
+                    await self.initialize_system()
+
+                response = self.agent.create_turn(
```
Comment (ashwinb) [inline examples/E2E-RAG-App/gradio_interface.py:145]: I think response is a mere Generator and not an AsyncGenerator. Maybe EventLogger is unnecessarily being an AsyncGenerator and that is causing this entire complexity around a thread. I will look into this, we need to get rid of all of this. This should look trivial like this:

```python
response = ...
for log in EventLogger().log(response): 
   yield history
```

Also I don't think "EventLogger" is the right abstraction at all (something for us to fix) -- you are reading things from the response, not logging!!

File: requirements.txt [hunk 16]
```diff
@@ -1,11 +1,11 @@
...
-httpx
 mesop
 python-dotenv
 setuptools
-uvicorn
 yfinance
-gradio
-llama-stack>=0.0.50
-llama-stack-client
+llama-stack==0.0.53rc1
```
Comment (ashwinb) [inline requirements.txt:6]: we certainly _cannot_ add all these dependencies just for a single app. we should put a separate set of requirements for the app and ask users to install them.

**Review by ashwinb — CHANGES_REQUESTED** [review]

File: examples/E2E-RAG-App/data/eval/eval.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+        {
+            "user_input": questions,
+            "response": generated_answers,
+            "retrieved_contexts": retrieved_contexts,
+            "reference": ground_truths,
+        }
+    )
+
+    # Run evaluation
+    result = evaluate(
```
Comment (yanxi0830) [inline examples/E2E-RAG-App/data/eval/eval.py:248]: We have native /evals API with Llama Stack now. Could we try evaluate the generated datasets with Llama Stack /evals & /scoring? Example here: https://github.com/meta-llama/llama-stack-apps/pull/117

File: examples/E2E-RAG-App/gradio_interface.py (hunk 15 above)
Comment (yanxi0830) [reply examples/E2E-RAG-App/gradio_interface.py:145]: Oh I think we can just make our Agent wrapper in our SDK to do sync CustomTool execution to make it return a Generator. https://github.com/meta-llama/llama-stack-client-python/blob/7495f517f1a35e82b78cc82378e439333dbe8641/src/llama_stack_client/lib/agents/agent.py#L79

Comment (ashwinb) [reply examples/E2E-RAG-App/gradio_interface.py:145]: oh interesting that it is tool execution that makes it async. in that case it is not necessarily a bad design since tool execution _will_ block for long amounts of time. I feel like maybe we need two methods `create_turn()` and `acreate_turn()` like most of the SDKs do because depending on the threading model of the caller, one is more natural vs. another.

Comment (heyjustinai) [reply examples/E2E-RAG-App/gradio_interface.py:145]: adding support for async workflows or separate create_turn() and acreate_turn() methods would be a very welcome change and simplify setups like these!

for context; I was initially a little puzzled as well but had similar realization as @ashwinb. so had to come out with the current implementation to bridge async functions with gradio’s synchronous event loop since gradio doesn’t support async generators directly. To enable streaming, I had to make the chat_stream function synchronous

File: examples/E2E-RAG-App/data/eval/eval.py (hunk 17 above)
Comment (heyjustinai) [reply examples/E2E-RAG-App/data/eval/eval.py:248]: That's amazing! will try it out

Comment (heyjustinai) [reply examples/E2E-RAG-App/data/eval/eval.py:248]: That's amazing! will take a look and try it out

File: examples/E2E-RAG-App/gradio_interface.py (hunk 15 above)
Comment (yanxi0830) [reply examples/E2E-RAG-App/gradio_interface.py:145]: https://github.com/meta-llama/llama-stack-client-python/pull/45 will make agent create_turn a sync function call. 
- apps migration: https://github.com/meta-llama/llama-stack-apps/pull/123

File: examples/E2E-RAG-App/gradio_interface.py (hunk 13 above)
Comment (heyjustinai) [reply examples/E2E-RAG-App/gradio_interface.py:118]: sounds good, will leave out "query_generator_config": {"type": "default", "sep": " "},

but since we are running ollama locally, we will need to keep the `max_tokens_in_context` and `max_chunks` for it to run at resonable speed

File: examples/E2E-RAG-App/gradio_interface.py (hunk 10 above)
Comment (wukaixingxp) [reply examples/E2E-RAG-App/gradio_interface.py:105]: The MODEL_NAME here is actually Ollama's model name which is difference than our meta MODEL_name, eg llama3.2:1b-instruct-fp16: "Llama3.2-1B-Instruct", I wonder if there is a easy way to do the model name mapping.

File: examples/E2E-RAG-App/gradio_interface.py (hunk 15 above)
Comment (heyjustinai) [reply examples/E2E-RAG-App/gradio_interface.py:142]: fixed to reflect this new change

File: requirements.txt (hunk 16 above)
Comment (heyjustinai) [reply requirements.txt:6]: agreed, creating a separate requirement.txt in project dir to reflect this @wukaixingxp  @init27

File: examples/E2E-RAG-App/gradio_interface.py (hunk 9 above)
Comment (heyjustinai) [reply examples/E2E-RAG-App/gradio_interface.py:71]: let me clarify the workflow:

1. Pre-ingestion (01_ingest_files.py):
   - converts various document formats (PDF, DOCX, etc.) into standardized Markdown files
   - also run vision model on images if present in docs
   - makes documents more easily parsable and consistent
   - only needs to be run once when adding new documents

2. Memory Bank Setup (in chat app):
   - takes the pre-processed Markdown files
   - creates vector embeddings and stores them in ChromaDB
   - this is a one-time setup when initializing the chat interface, after setup, the chat app uses this existing memory bank for retrievals

So you're correct - the pre-ingestion step (01_ingest_files.py) handles document standardization and vision hanlding, while the memory bank setup handles vectorization and storage.

File: examples/E2E-RAG-App/01_ingest_files.py (hunk 2 above)
Comment (heyjustinai) [reply examples/E2E-RAG-App/01_ingest_files.py]: fixed

File: examples/E2E-RAG-App/README.md (hunk 5 above)
Comment (wukaixingxp) [reply examples/E2E-RAG-App/README.md]: `llama_stack_start.sh`  is the docker startup script that will be run inside of docker. run_RAG.sh will create the docker.

File: examples/E2E-RAG-App/README.md (hunk 4 above)
Comment (wukaixingxp) [reply examples/E2E-RAG-App/README.md]: Sure.. I will make a short version

File: examples/E2E-RAG-App/README.md (hunk 3 above)
Comment (wukaixingxp) [reply examples/E2E-RAG-App/README.md]: Sure.. I will make a short version

File: examples/DocQA/docker/chroma_start.sh [hunk 18]
```diff
@@ -0,0 +1,3 @@
+pip install chromadb
```
Comment (ashwinb) [inline examples/DocQA/docker/chroma_start.sh:1]: why is this needed? if you are starting from a chroma image you don't need it I think.

also can we inline this script into docker compose?

File: examples/DocQA/docker/llama_stack_start.sh [hunk 19]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# Sleep for 45 seconds
+sleep 45
```
Comment (ashwinb) [inline examples/DocQA/docker/llama_stack_start.sh:3]: why is this needed?!

File: examples/DocQA/app.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+# Load environment variables
+load_dotenv()
+
+HOST = os.getenv("HOST", "localhost")
+PORT = int(os.getenv("PORT", "5000"))
+GRADIO_SERVER_PORT = int(os.getenv("GRADIO_SERVER_PORT", "7861"))
+USE_GPU = os.getenv("USE_GPU", False)
+MODEL_NAME = os.getenv("MODEL_NAME", "meta-llama/Llama-3.2-1B-Instruct")
+# if use_gpu, then the documents will be processed to output folder
+DOCS_DIR = "/root/rag_data/output" if USE_GPU else "/root/rag_data/"
```
Comment (ashwinb) [inline examples/DocQA/app.py:31]: why is this different depending on whether we use a GPU or not?

File: examples/DocQA/app.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+from llama_stack_client.types.agent_create_params import AgentConfig
+from llama_stack_client.types.memory_insert_params import Document
+import re
+
+
+# Load environment variables
+load_dotenv()
+
+HOST = os.getenv("HOST", "localhost")
+PORT = int(os.getenv("PORT", "5000"))
```
Comment (ashwinb) [inline examples/DocQA/app.py:26]: I think this should be `LLAMA_STACK_PORT` to make it clearer

File: examples/DocQA/app.py [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+GRADIO_SERVER_PORT = int(os.getenv("GRADIO_SERVER_PORT", "7861"))
+USE_GPU = os.getenv("USE_GPU", False)
```
Comment (ashwinb) [inline examples/DocQA/app.py:28]: I think this is really `USE_GPU_FOR_DOC_INGESTION`

File: examples/DocQA/docker/llama_stack_start.sh [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+# Print a message indicating the start of llama-stack docker
```
Comment (ashwinb) [inline examples/DocQA/docker/llama_stack_start.sh:4]: the comments in this script are a lot of noise. they don't help.

File: examples/DocQA/docker/llama_stack_start.sh [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+# Check if GPU is enabled and run ingest files script accordingly
+if [ "$USE_GPU" = true ]; then
+  echo "Using GPU to ingest files"
+  pip install docling
+  python /root/DocQA/scripts/ingest_files.py --input_dir /root/rag_data/
+fi
+# Print a message indicating the start of llama-stack server
+echo "starting the llama-stack server"
+# Run llama-stack server with specified config and disable ipv6
+python -m llama_stack.distribution.server.server --yaml-config /root/my-run.yaml --disable-ipv6&
```
Comment (ashwinb) [inline examples/DocQA/docker/llama_stack_start.sh:18]: I think this should be re-architected a bit. Why isn't the docker running two services for this purpose:
- one for running the llama stack server
- one for running the RAG app? that entrypoint can just be `python /root/DocQA/app.py`

File: examples/DocQA/docker/run_RAG.sh [hunk 25]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+echo "DOC_PATH=$(jq -r '.document_path' ./RAG_service.json)" > .env
```
Comment (ashwinb) [inline examples/DocQA/docker/run_RAG.sh:3]: why do we need a .json and an .env? why can't we just use .env and avoid this manipulation completely?

**Review by ashwinb — CHANGES_REQUESTED** [review]

File: examples/DocQA/docker/chroma_start.sh (hunk 18 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/chroma_start.sh:1]: Sure.. this has been changed

File: examples/DocQA/docker/llama_stack_start.sh (hunk 19 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/llama_stack_start.sh:3]: Ollama needs to download the model for the first time and llama-stack will produce ollama can not found if we do not wait for ollama to be complete. I borrow the idea from [here](https://github.com/meta-llama/llama-stack/blob/435f34b05e84f1747b28570234f25878cf0b31c4/distributions/dell-tgi/compose.yaml#L45)

File: examples/DocQA/docker/llama_stack_start.sh (hunk 23 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/llama_stack_start.sh:4]: sure, it has been changed

File: examples/DocQA/docker/llama_stack_start.sh (hunk 24 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/llama_stack_start.sh:18]: I was hoping to use `LlamaStackDirectClient `, but I am not sure if `LlamaStackDirectClient` supports MemoryBank connection to ChromaDB.

File: examples/DocQA/docker/run_RAG.sh (hunk 25 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/run_RAG.sh:3]: Thanks, this has been changed

File: examples/DocQA/app.py (hunk 21 above)
Comment (wukaixingxp) [reply examples/DocQA/app.py:26]: Sure, this has been changed

File: examples/DocQA/app.py (hunk 20 above)
Comment (wukaixingxp) [reply examples/DocQA/app.py:31]: Our ingest pipeline will take the `/root/rag_data/` data folder and save results to `/root/rag_data/output` when using GPU. Otherwise the data folder is just `/root/rag_data/`

Comment (ashwinb) [reply examples/DocQA/app.py]: @wukaixingxp still don't know why the output folder is _different_ when using GPU vs. not? if you aren't using GPU, where does the ingest pipeline save the results?

File: examples/DocQA/docker/llama_stack_start.sh [hunk 26]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+sleep 45
+echo "-----starting to llama-stack docker now---------"
+pip install gradio
+
+if [ "$USE_GPU_FOR_DOC_INGESTION" = true ]; then
```
Comment (ashwinb) [inline examples/DocQA/docker/llama_stack_start.sh:7]: looks like if that variable is false, we just don't ingest at all?

File: examples/DocQA/docker/llama_stack_start.sh (hunk 19 above)
Comment (ashwinb) [reply examples/DocQA/docker/llama_stack_start.sh:3]: ok we should add a comment there saying this is a wait for the model to get loaded

File: examples/DocQA/docker/llama_stack_start.sh (hunk 26 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/llama_stack_start.sh:7]: Yes, ingest pipeline is just for image: basically it takes a PDF, find out if there is any image and use 11B to generate image description, then it will save the original text with image description into `output` folder so that everything is now text data, ready to be used by 3B RAG agent. Running 11B on CPU is sooo slow and we have not enable this in the current stage. We believe this can be a P1 feature to have.

File: examples/DocQA/app.py (hunk 20 above)
Comment (wukaixingxp) [reply examples/DocQA/app.py]: ingest pipeline is just for image: basically it takes everything in `/root/rag_data`, find out if there is any image embed in the PDF, split the image and use 11B to generate image description, then it will save the original text with image description into `/root/rag_data` folder so that everything is now text data, ready to be used by 3B RAG agent. Running 11B on CPU is sooo slow and we have not enable this in the current stage. We believe this can be a P1 feature to have. Now we only support text data and will ignore embedded images, thus just take everything in the `/root/rag_data` folder

File: examples/DocQA/docker/llama_stack_start.sh (hunk 19 above)
Comment (wukaixingxp) [reply examples/DocQA/docker/llama_stack_start.sh:3]: sure, will add this

**Comment by ashwinb** [comment]
Can you rebase and update `requirements.txt` so we can merge it in
